Require `belongs_to` by default.

simi opened the pull request:
Deprecate `required` option in favor of `optional` for `belongs_to`.

closes #18233

simi committed:
Require `belongs_to` by default.

Deprecate `required` option in favor of `optional` for belongs_to.

simi commented:
/cc @dhh @al2o3cr

simi commented:
I wonder what to do with failing test.

```
rails generate scaffold LineItems product:references cart:belongs_to
bundle exec rake db:migrate test`
```

This code fails now, since belongs_to relation is required by default. Should I change generated test for belongs_to to make it pass or should I remove this test since it is not valid with this change?

simi commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -110,5 +110,24 @@ def self.add_destroy_callbacks(model, reflection)
+
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        ActiveSupport::Deprecation.warn('`required` option for belongs_to relation is deprecated and replaced with `optional` option.')
```
Better info will be needed here. Any ideas?

dhh replied:
'`required: true` is now the default for belongs_to and can be removed.' – don't even need to talk about optional here. What they wanted to do is now being done for them.

simi replied:
What about when `required: false` is specified explicitly? Maybe throw deprecation warning only for `false` value?

dhh replied:
That was the old default, which means no client code will have that code, since it doesn’t make any sense.

> On Feb 15, 2015, at 12:54 PM, Josef Šimánek <notifications@github.com> wrote:
> 
> In activerecord/lib/active_record/associations/builder/belongs_to.rb <https://github.com/rails/rails/pull/18937#discussion_r24725940>:
> 
> > @@ -110,5 +110,24 @@ def self.add_destroy_callbacks(model, reflection)
> >        name = reflection.name
> >        model.after_destroy lambda { |o| o.association(name).handle_dependency }
> >      end
> > +
> > +    def self.define_validations(model, reflection)
> > +      if reflection.options.key?(:required)
> > +        ActiveSupport::Deprecation.warn('`required` option for belongs_to relation is deprecated and replaced with `optional` option.')
> What about when required: false is specified explicitly? Maybe throw deprecation warning only for false value?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/18937/files#r24725940>.
>

simi replied:
OK, makes sense.

dhh commented:
You can kill that test and replace it with a test that tests optional: true.

> On Feb 14, 2015, at 3:24 AM, Josef Šimánek <notifications@github.com> wrote:
> 
> I wonder what to do with failing test.
> 
> rails generate scaffold LineItems product:references cart:belongs_to
> bundle exec rake db:migrate test`
> This code fails now, since belongs_to relation is required by default. Should I change generated test for belongs_to to make it pass or should I remove this test since it is not valid with this change?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/18937#issuecomment-74371400>.
>

simi commented:
Should I add also this to generator? `rails g scaffold post author:belongs_to:optional` ?

dhh commented:
Nah, don’t think that’s necessary. Will be a very uncommon option to use, I think.

> On Feb 18, 2015, at 14:30, Josef Šimánek <notifications@github.com> wrote:
> 
> Should I add also this to generator? rails g scaffold post author:belongs_to:optional ?
> 
> —
> Reply to this email directly or view it on GitHub.
>

simi commented:
OK I think my work is almost done here. I'll add also CHANGELOG note soon.

dhh commented:
Awesome. Ping when ready to merge.

> On Feb 18, 2015, at 15:35, Josef Šimánek <notifications@github.com> wrote:
> 
> OK I think my work is almost done here. I'll add also CHANGELOG note soon.
> 
> —
> Reply to this email directly or view it on GitHub.
>

simi commented:
I added CHANGELOG note, but it is a little fuzzy to me. Feel free to ping me with better one.

dhh commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+    This can by configure by `config.active_record.require_belongs_to` option
+    or by `optional: true` option on relation.
+
+    *Josef Šimánek*
```
Let's try with:

`belongs_to` will now trigger a validation error by default if the association is not present. You can turn this off on a per-association basis with `optional: true`. (Note this new default only applies to new Rails apps that will be generated with `config.active_record.require_belongs_to = true`)

simi replied:
Actually the final note is not true right now. `config.active_record.require_belongs_to = true` is set directly in railtie, not in application config. Should I move it to new generated `application.rb` and let it `nil` for existing applications? In that case your note will be true.

dhh commented:
Let's update the config option to be a bit clearer: `belongs_to_required_by_default = true`

dhh commented:
This is also missing the config/initializers/active_record_belongs_to_required_by_default.rb file that's going to set the `belongs_to_required_by_default = true` setting by default for new apps.

simi commented:
@dhh I mentioned that https://github.com/rails/rails/pull/18937/files#r24953865 already.

dhh commented:
Ah, we crossed streams. Didn't note that one. Yes, please do go ahead. Thanks!

jeremy commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -110,5 +110,23 @@ def self.add_destroy_callbacks(model, reflection)
+        required = model.require_belongs_to
+      else
+        required = !reflection.options[:optional]
+      end
```
I'm finding the control flow here a little difficult to follow. Consider condensing the conditionals and using the `:optional` option to determine whether to declare the presence validation:
```ruby
# Disallow conflicting options
if reflection.options.key?(:required) && reflection.options.key?(:optional)
  raise ArgumentError, "complain if we have conflicting options"
# Deprecate :required
elsif reflection.options.key?(:required)
  reflection.options[:optional] = !reflection.options[:required]
  ActiveSupport::Deprecation.warn "switch to `optional: #{reflection.options[:optional]}`"
# Set default :optional option
elsif !reflection.options.key?(:optional)
  reflection.options[:optional] = !model.require_belongs_to
end

super

if !reflection.options[:optional]
  model.validates_presence_of …
end
```

simi replied:
@jeremy I try to rethink this flow, but deprecation was marked as unnecessary by @dhh here.

jeremy commented on activerecord/lib/active_record/associations/builder/has_one.rb:
```diff
@@ -17,5 +17,12 @@ def self.valid_dependent_options
+      if reflection.options[:required]
+        model.validates_presence_of reflection.name, message: :required
+      end
+    end
```
I think it'll be odd to switch to `optional:` on belongs_to but keep `required:` for has_one.

dhh replied:
I don't think so. belongs_to to me strongly implies a owner/ownee relationship. Doesn't make sense in most cases to say I'm an ownee without a owner. But the other case does make sense. That the owner may or may not have an ownee. (Excuse the made-up word).

jeremy commented on activerecord/lib/active_record/core.rb:
```diff
@@ -87,6 +87,8 @@ def self.configurations
       mattr_accessor :maintain_test_schema, instance_accessor: false
 
+      mattr_accessor :require_belongs_to, instance_accessor: false
+
```
Out of context, I wouldn't know what this "require belongs to" means. Is there a clearer config name?

simi replied:
yup, it is on the way

jeremy commented on activerecord/test/cases/associations/required_test.rb:
```diff
@@ -34,7 +34,7 @@ class Child < ActiveRecord::Base
   test "required belongs_to associations have presence validated" do
     model = subclass_of(Child) do
-      belongs_to :parent, required: true, inverse_of: false,
+      belongs_to :parent, optional: false, inverse_of: false,
```
If we still support `required: true`, we should still test it.

jeremy commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -32,6 +32,7 @@ class Railtie < Rails::Railtie # :nodoc:
 
     config.active_record.use_schema_cache_dump = true
     config.active_record.maintain_test_schema = true
+    config.active_record.require_belongs_to = true
```
Since this is in the Railtie, it'll surprisingly affect apps that upgrade to newer Rails. We'll need to generate new apps with it set, but use `false` as the default to preserve existing app behavior.

simi replied:
sure, it is on the way also

jeremy commented:
Nice work @simi. Pardon the repeated remarks, looks like I missed concurrent feedback too. Thanks!

simi commented:
Nevermind @jeremy. Thanks for review.

I have just one more question. I made change where `config.active_record.require_belongs_to = true` is generated to `application.rb` for new apps. But it is only one uncommented (unless :skip_active_record) configuration in that file. Should I left it there or should it be better to move into custom initializer?

simi commented:
I bet it will be better as initializer after inspecting current generated initializers.

simi commented:
@jeremy @dhh I pushed new commit (I can squash later, just to highlight new code). `config.active_record.belongs_to_required_by_default = true` is now part of `config/initializers/active_record_belongs_to_required_by_default.rb`.

This initializer is generated only when `ActiveRecord` is not skipped and is not added for `rake rails:update`. I think this is compatible enough to skip deprecation.

dhh commented:
I think we want the deprecation still in place because we want people to switch over. That is, set config.active_record.belongs_to_required_by_default themselves, and then update with optional or nothing as needed. Because we'll remove config.active_record.belongs_to_required_by_default in Rails 5.x or 6.

simi commented:
Ahh, ok. Probably I misunderstood you in comments. I'll add deprecation back.

simi commented:
I mean here - https://github.com/rails/rails/pull/18937#discussion-diff-24725280.

dhh commented:
Hmm, actually. Good point. Not sure there is a good way to inform about this via a deprecation. I think we’ll have to simply make the deprecation a matter of documentation on the belongs_to method.

> On Feb 18, 2015, at 5:57 PM, Josef Šimánek <notifications@github.com> wrote:
> 
> I mean here - #18937 (diff) <https://github.com/rails/rails/pull/18937#discussion-diff-24725280>.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/18937#issuecomment-74986900>.
>

simi commented:
I'll add this into documentation anyway. 

Deprecation can be returned also to https://github.com/rails/rails/pull/18937/files#diff-a26dc1fd56f9dac74db1300aa767378dR24.

dhh commented:
I think all we need is the documentation updates now.

> On Feb 18, 2015, at 6:03 PM, Josef Šimánek <notifications@github.com> wrote:
> 
> I'll add this into documentation anyway.
> 
> Deprecation can be returned also to https://github.com/rails/rails/pull/18937/files#diff-a26dc1fd56f9dac74db1300aa767378dR24 <https://github.com/rails/rails/pull/18937/files#diff-a26dc1fd56f9dac74db1300aa767378dR24>.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/18937#issuecomment-74987430>.
>

simi commented:
I added some really basic documentation note.

![screenshot-localhost 3333 2015-02-19 03-17-43](https://cloud.githubusercontent.com/assets/193936/6260922/dde41da0-b7e6-11e4-915c-9eba00c1ddde.png)

I did a quick read thru guides and I changed validated documentation to optional for belongs_to there.

dhh commented:
Looks good. We can add further refinements later!

> On Feb 18, 2015, at 6:31 PM, Josef Šimánek <notifications@github.com> wrote:
> 
> I added some really basic documentation note.
> 
>  <https://cloud.githubusercontent.com/assets/193936/6260922/dde41da0-b7e6-11e4-915c-9eba00c1ddde.png>
> I did a quick read thru guides and I changed validated documentation to optional for belongs_to there.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/18937#issuecomment-74989664>.
>

simi commented:
@jeremy what about https://github.com/rails/rails/pull/18937/files#diff-d853d726a0a949328b3a7fb8f9592c7eR114? Are you still considering it hard to follow?

rafaelfranca commented on guides/source/association_basics.md:
```diff
@@ -832,7 +832,7 @@ The `belongs_to` association supports these options:
 * `:polymorphic`
 * `:touch`
-* `:validate`
+* `:optional`
```
We still have the `:validate` option. It is not the same as `:required` so we should keep its documentation.

rafaelfranca commented:
We need to document `belongs_to_required_by_default` in the configuration guide.

Also had you tested the initializer in a new application? I'm not sure if it will work as it is right now because of load order of Active Record.

simi commented:
@rafaelfranca I reverted my `validate` doc changes and added a note into configuration guide.

simi commented:
And I tested this via :

```
bin/rails new artest --dev
cd artest
bin/rails g model post
bin/rails g model comment post:belongs_to
bin/rake db:migrate
bin/rails runner "p Comment.new.valid?" #=> false
RAILS_ENV=production bin/rake db:migrate
RAILS_ENV=production bin/rails runner "p Comment.new.valid?" #=> false

```

mockdeep commented:
Just came across this pull request. I think validating presence of `belongs_to` relationships is going to result in a lot of unnecessary additional queries in codebases, especially if you're validating a number of records before bulk importing or updating them in the database. We've been moving in the direction of instead validating presence of the associated `_id` and letting the foreign key constraint in the database ensure the associated record exists.

Silex commented:
I feel this change was a bit of a mistake at the design level... validations and relations are different things, and now we get validations done when defining relations.

I understand the "it's convenient" argument tho.

Silex commented on guides/source/association_basics.md:
```diff
@@ -956,6 +957,10 @@ end
 
+##### `:optional`
+
+If you set the `:optional` option to `true`, then associated object will be validated for presence. By default, this is `false`: associated objects will be validated for presence.
```
Typo: "will NOT be validated for presence"

simi replied:
**optional:** By default, this is **false**: associated objects will **be** validated for presence.

Silex replied:
Look more closely, the sentence says the same thing twice:

```
If you set the `:optional` option to `true`, then associated object 
will be validated for presence. 
By default, this is `false`: associated objects 
will be validated for presence.
```

I took the original sentence and just modified the alignement so it's visually clear they are the same.

Silex replied:
Ok, this was fixed in https://github.com/rails/rails/commit/803ef74f9bbcbc71af12060157a996f7b65e24db

lekansogunle commented:
We are only testing for validation of the presence of the association ID, what if the given ID is invalid. And app behaviour is not to validate the existence of such association?
